src/lib.rs

```diff
@@ -19,139 +19,76 @@
+//! - the ustatus module which provices uProtocol types for representing status and status codes
+//! - the utransport module as an interface contract between uProtocol and specific transport protocol implementations
+//! - the uuid module which generates and validates UUIDs as per the uProtocol specification
+//!
+//! For user convenience, all of these modules export their types on up_rust top-level, except for (future) optional features.
 //!
 //! ## References
 //! - [Eclipse-uProtocol Specification](https://github.com/eclipse-uprotocol/up-spec)
 //! - [Eclipse-uProtocol Core API types](https://github.com/eclipse-uprotocol/up-core-api)
 
-mod types {
-    pub mod parsingerror;
-    pub mod serializationerror;
-    pub mod validationerror;
-}
-
-pub mod cloudevent {
-    pub mod builder {
-        mod ucloudeventbuilder;
-        mod ucloudeventutils;
-
-        pub use ucloudeventbuilder::*;
-        pub use ucloudeventutils::*;
-    }
-    pub mod datamodel {
-        mod ucloudeventattributes;
-
-        pub use ucloudeventattributes::*;
-    }
-    pub mod serializer {
-        mod cloudeventjsonserializer;
-        mod cloudeventprotobufserializer;
-        mod cloudeventserializer;
+// up_core_api types used and augmented by up_rust - symbols re-exported to toplevel, errors are module-specific
+mod rpc;
+pub use rpc::{CallOptions, CallOptionsBuilder};
+pub use rpc::{RpcClient, RpcClientResult};
+pub use rpc::{RpcMapper, RpcMapperError};
+pub use rpc::{RpcPayload, RpcPayloadResult, RpcResult};
+pub use rpc::{RpcServer, RpcServerListener};
 
-        pub use crate::types::serializationerror::*;
-        pub use cloudeventjsonserializer::*;
-        pub use cloudeventprotobufserializer::*;
-        pub use cloudeventserializer::*;
-    }
-    pub mod validator {
-        mod cloudeventvalidator;
+mod uattributes;
+pub use uattributes::{
+    PublishValidator, RequestValidator, ResponseValidator, UAttributesValidator, Validators,
+};
+pub use uattributes::{UAttributes, UMessageType, UPriority};
+pub use uattributes::{UAttributesParsingError, UAttributesValidationError};
 
-        pub use crate::types::validationerror::*;
-        pub use cloudeventvalidator::*;
-    }
-}
+mod umessage;
+pub use umessage::{UMessage, UMessageBuilder, UMessageBuilderError};
 
-// protoc-generated stubs, see build.rs
-include!(concat!(env!("OUT_DIR"), "/cloudevents/mod.rs"));
+mod upayload;
+pub use upayload::{Data, UPayload, UPayloadFormat};
+pub use upayload::{UPayloadParsingError, UPayloadSerializationError};
 
-pub mod rpc {
-    mod calloptions;
-    mod rpcclient;
-    mod rpcmapper;
-    mod rpcresult;
-    mod rpcserver;
+mod uri;
+pub use uri::{AddressType, Number};
+pub use uri::{UAuthority, UEntity, UResource, UResourceBuilder};
+pub use uri::{UUri, UriValidator};
+pub use uri::{UriParsingError, UriSerializationError, UriValidationError};
```

Comment:
I like that we call these `UriParsingError` etc instead of `UUriparsingError`.
Can we do the same with the names of the other error types, i.e. use `AttributesParsingError` instead of `UAttributesParsingError` and `PayloadSerializationError` instead of `UPayloadSerializationError`? Having the names prefixed by the `U` feels a little strange to me and we should at least use a consistent naming strategy ...

Reply:
I do share the sentiment that the leading 'U's are awkward - at the same time, it is very hard to draw a consistent line where to have vs where to omit it. So I'm picking up this input, but will go the opposite way - and rather re-introduce the leading-U to the Uri types, to achieve overall consistency.